### PR TITLE
Add builders to elements.

### PR DESCRIPTION
**Pull request by JakeWharton**
Next step: move validation to happen on every `addFoo` method call for more local problem detection.

### CODE REVIEW

**Comment by JakeWharton** [comment]
Sorry this is long... mostly mechanical. Very JavaPoet-esque API now. This will make changes in the future easier (e.g., adding `oneof` would have been really nice _after_ this change).

File: src/main/java/com/squareup/protoparser/EnumConstantElement.java [hunk 1]
```diff
@@ -47,4 +45,38 @@ public static EnumConstantElement create(String name, int tag, String documentat
...
+  public static final class Builder {
+    private String name;
+    private Integer tag;
+    private String documentation = "";
+    private final List<OptionElement> options = new ArrayList<>();
+
+    private Builder() {
+    }
+
+    public Builder setName(String name) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/protoparser/EnumConstantElement.java]: Need the `set` prefix? Effective Java 2e doesn't. I slightly don't like it, but recognize that I'm odd-man out

File: src/main/java/com/squareup/protoparser/EnumElement.java [hunk 2]
```diff
@@ -103,4 +98,51 @@ public static EnumElement create(String name, String qualifiedName, String docum
...
+    public Builder addOption(OptionElement option) {
+      options.add(checkNotNull(option, "option"));
+      return this;
+    }
+
+    public EnumElement build() {
+      if (!parseAllowAlias(options)) {
+        validateTagUniqueness(qualifiedName, constants);
+      }
+      return new AutoValue_EnumElement(name, qualifiedName, documentation,
```
Comment (swankjesse) [inline src/main/java/com/squareup/protoparser/EnumElement.java]: ooooh

File: src/main/java/com/squareup/protoparser/ProtoParser.java [hunk 3]
```diff
@@ -193,14 +195,15 @@ private Object readDeclaration(String documentation, Context context) {
...
-    String previousPrefix = prefix;
     String name = readName();
+    MessageElement.Builder builder = MessageElement.builder()
+        .setName(name)
+        .setQualifiedName(prefix + name)
+        .setDocumentation(documentation);
+
+    String previousPrefix = prefix;
     prefix = prefix + name + ".";
-    List<FieldElement> fields = new ArrayList<>();
```
Comment (swankjesse) [inline src/main/java/com/squareup/protoparser/ProtoParser.java]: Sweet.

File: src/test/java/com/squareup/protoparser/ProtoParserTest.java [hunk 4]
```diff
@@ -263,15 +280,33 @@
...
-    TypeElement expected = MessageElement.create("SearchRequest", "SearchRequest", "", list(
-            FieldElement.create(REQUIRED, "string", "query", 1, "", NO_OPTIONS)),
-        list(OneOfElement.create("page_info", "",
-            list(FieldElement.create(ONE_OF, "int32", "page_number", 2, "", NO_OPTIONS),
-                FieldElement.create(ONE_OF, "int32", "result_per_page", 3, "", NO_OPTIONS)))),
-        NO_TYPES, NO_EXTENSIONS, NO_OPTIONS);
-    ProtoFile protoFile = ProtoFile.builder("search.proto").addType(expected).build();
-    assertThat(ProtoParser.parse("search.proto", new StringReader(proto))) //
-        .isEqualTo(protoFile);
+    ProtoFile expected = ProtoFile.builder("search.proto")
```
Comment (swankjesse) [inline src/test/java/com/squareup/protoparser/ProtoParserTest.java]: So much more legible

**Comment by swankjesse** [comment]
Huge win. More legible, more predictable, and you don't need to pass in nulls and empty strings when you don't care.

LGTM.

**Pull request merged**